client/mysql.cc

```diff
@@ -1116,6 +1116,24 @@ inline int get_command_index(char cmd_char)
   return -1;
 }
 
+static void redact_mysql_pwd(void)
+{
+#ifndef DONT_USE_MYSQL_PWD
+  /* defense-in-depth: overwrite password in original environment vector */
+  char *passwd = getenv("MYSQL_PWD");
+  if (passwd) {
+      /* setenv copies passwd, result not visible in /proc/$pid/environ */
+      if (setenv("MYSQL_PWD", passwd, 1) == -1) {
+        put_info("setenv() failed", INFO_ERROR, 0);
+        my_end(0);
+        exit(1);
+      }
+      /* overwrite password in /proc/$pid/environ */
+      memset(passwd, 'x', strlen(passwd));
```

Comment:
is there a way to overwrite this without exposing the length of the password? Does this work on Windows/BSD?

Reply:
I don't know a method where the length isn't exposed. FWIW, when passing the password in the argument vector (with -p/--password) the length is exposed, as well. I don't know if this is also effective on Windows/BSD. I don't even know if those systems expose the environments variables to other processes in a similar way.